{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Finding differentialy expressed miRNA's between LUAD stages I, II, III, IV vs normal using Logit regression\n",
    "### Utilize Group Lasso with MISIM miRNA similarity data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# Necessary imports\n",
    "%load_ext autoreload\n",
    "%autoreload 2\\\n",
    "    \n",
    "import os\n",
    "import pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from definitions import ROOT_DIR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load normal and cancer miRNA expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mirna_tumor_df.shape (513, 1882) , nulls: 0\nmirna_normal_df.shape (46, 1882) , nulls: 0\nvalidated_miRNAs.shape (34, 2)\nnormal    46\nName: pathologic_stage, dtype: int64\nStage I      277\nStage II     121\nStage III     84\nStage IV      24\nName: pathologic_stage, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "from definitions import ROOT_DIR\n",
    "\n",
    "# Load files into pandas data frames\n",
    "mirna_tumor_df = pandas.read_csv(os.path.join(ROOT_DIR, \"data/processed/miRNA/tumor_miRNA.csv\"))\n",
    "mirna_normal_df = pandas.read_csv(os.path.join(ROOT_DIR, \"data/processed/miRNA/normal_miRNA.csv\"))\n",
    "clinical_df = pandas.read_csv(os.path.join(ROOT_DIR, \"data/processed/clinical/clinical.csv\"))\n",
    "validated_miRNA_csv = pandas.read_csv(os.path.join(ROOT_DIR, 'data/external/validated_luad_miRNAs_miRCancer.csv'))\n",
    "\n",
    "# Print data frame shapes\n",
    "print \"mirna_tumor_df.shape\", mirna_tumor_df.shape, ', nulls:', mirna_tumor_df.isnull().sum().sum()\n",
    "print \"mirna_normal_df.shape\", mirna_normal_df.shape, ', nulls:', mirna_normal_df.isnull().sum().sum()\n",
    "print 'validated_miRNAs.shape', validated_miRNA_csv.shape\n",
    "\n",
    "# Merge normal and tumor miRNA expression profiles with clinical cancer stage data\n",
    "mirna_normal = pandas.merge(clinical_df[['patient_barcode', 'pathologic_stage']], mirna_normal_df, on='patient_barcode')\n",
    "mirna_normal['pathologic_stage'] = 'normal'\n",
    "mirna_tumor = pandas.merge(clinical_df[['patient_barcode', 'pathologic_stage']], mirna_tumor_df, on='patient_barcode')\n",
    "\n",
    "mirna_tumor.dropna(inplace=True)\n",
    "mirna_normal.dropna(inplace=True)\n",
    "\n",
    "# Map stage IA to stage I, IB to I, etc. ...\n",
    "pathologic_stage_map = {'Stage IA': 'Stage I', 'Stage IB': 'Stage I', \n",
    "                        'Stage IIA': 'Stage II', 'Stage IIB': 'Stage II', \n",
    "                        'Stage IIIA': 'Stage III', 'Stage IIIB': 'Stage III'}\n",
    "mirna_tumor.replace({'pathologic_stage': pathologic_stage_map}, inplace=True)\n",
    "\n",
    "# Store list of all miRNA's in miRNA expression data\n",
    "mirna_list = list(mirna_tumor.columns)[2:]\n",
    "\n",
    "# Print quick overview of data\n",
    "print mirna_normal['pathologic_stage'].value_counts().sort_index(axis=0)\n",
    "print mirna_tumor['pathologic_stage'].value_counts().sort_index(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load normal & cancer Gene Expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20530, 517)\n(20530, 61)\n"
     ]
    }
   ],
   "source": [
    "gene_exp_tumor_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/processed/gene_expression/tumor/READ__illuminahiseq_rnaseqv2__GeneExp.txt'), \n",
    "                                      header=0, delimiter='\\t')\n",
    "gene_exp_normal_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/processed/gene_expression/normal/READ__illuminahiseq_rnaseqv2__GeneExp.txt'), \n",
    "                                      header=0, delimiter='\\t')\n",
    "\n",
    "gene_exp_tumor_df.rename(columns=lambda x: x[:12], inplace=True)\n",
    "gene_exp_normal_df.rename(columns=lambda x: x[:12], inplace=True)\n",
    "\n",
    "print gene_exp_tumor_df.shape\n",
    "print gene_exp_normal_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reshape gene expression data frames to have shape (patients x gene expression) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_symbols 20502\ngene_exp_tumor_patients 515\ngene_exp_normal_patients 59\n(515, 20503)\n(59, 20503)\n"
     ]
    }
   ],
   "source": [
    "# Remove entries with unknown Gene Symbol\n",
    "gene_exp_tumor_df = gene_exp_tumor_df[gene_exp_tumor_df.GeneSymbol != '?']\n",
    "gene_exp_normal_df = gene_exp_normal_df[gene_exp_normal_df.GeneSymbol != '?']\n",
    "\n",
    "# Get list of all gene_symbols\n",
    "gene_symbols = list(gene_exp_tumor_df['GeneSymbol'])\n",
    "# Get list of tumor and normal patient_barcode\n",
    "gene_exp_tumor_patient_barcodes = list(gene_exp_tumor_df.columns)[2:]\n",
    "gene_exp_normal_patient_barcodes = list(gene_exp_normal_df.columns)[2:]\n",
    "\n",
    "# Drop EntrezID column\n",
    "gene_exp_tumor = gene_exp_tumor_df.drop(['EntrezID', 'GeneSymbol'], axis=1)\n",
    "gene_exp_normal = gene_exp_normal_df.drop(['EntrezID', 'GeneSymbol'], axis=1)\n",
    "\n",
    "# Reshaping data frame to have columns for GeneSymbols, and rows of patients\n",
    "gene_exp_tumor = gene_exp_tumor.T\n",
    "gene_exp_normal = gene_exp_normal.T\n",
    "gene_exp_tumor.columns = gene_symbols\n",
    "gene_exp_normal.columns = gene_symbols\n",
    "\n",
    "# Add column for patients barcode\n",
    "gene_exp_tumor['patient_barcode'] = gene_exp_tumor.index\n",
    "gene_exp_normal['patient_barcode'] = gene_exp_normal.index\n",
    "\n",
    "print \"gene_symbols\", len(gene_symbols)\n",
    "print \"gene_exp_tumor_patients\", len(gene_exp_tumor_patient_barcodes)\n",
    "print \"gene_exp_normal_patients\", len(gene_exp_normal_patient_barcodes)\n",
    "\n",
    "print gene_exp_tumor.shape\n",
    "print gene_exp_normal.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter samples with matched Gene Expression data and miRNA data\n",
    "## More data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stage I      275\nStage II     120\nStage III     84\nStage IV      24\nName: pathologic_stage, dtype: int64\nnormal    20\nName: pathologic_stage, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:26: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame\n\nSee the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n/opt/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame\n\nSee the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n/opt/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:28: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame\n\nSee the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n/opt/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:29: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame\n\nSee the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merged_mirna_tumor (503, 1881) nulls: 0\nmerged_mirna_normal (20, 1881) nulls: 0\nmerged_gene_exp_tumor (503, 20484) nulls: 0\nmerged_gene_exp_normal (20, 20484) nulls: 0\n"
     ]
    }
   ],
   "source": [
    "# Merge normal and tumor miRNA expression profiles with clinical cancer stage data\n",
    "merged_tumor_patients = pandas.merge(gene_exp_tumor[['patient_barcode']], mirna_tumor, on='patient_barcode')[['patient_barcode', 'pathologic_stage']]\n",
    "merged_normal_patients = pandas.merge(gene_exp_normal[['patient_barcode']], mirna_normal, on='patient_barcode')[['patient_barcode', 'pathologic_stage']]\n",
    "\n",
    "# Print quick overview of data\n",
    "print merged_tumor_patients['pathologic_stage'].value_counts(dropna=False).sort_index(axis=0)\n",
    "print merged_normal_patients['pathologic_stage'].value_counts(dropna=False).sort_index(axis=0)\n",
    "\n",
    "# Drop NA\n",
    "merged_tumor_patients.dropna(inplace=True)\n",
    "merged_normal_patients.dropna(inplace=True)\n",
    "\n",
    "# Filter samples\n",
    "merged_mirna_tumor = mirna_tumor[mirna_tumor['patient_barcode'].isin(merged_tumor_patients['patient_barcode'])]\n",
    "merged_mirna_normal = mirna_normal[mirna_normal['patient_barcode'].isin(merged_normal_patients['patient_barcode'])]\n",
    "merged_gene_exp_tumor = gene_exp_tumor[gene_exp_tumor['patient_barcode'].isin(merged_tumor_patients['patient_barcode'])]\n",
    "merged_gene_exp_normal = gene_exp_normal[gene_exp_normal['patient_barcode'].isin(merged_normal_patients['patient_barcode'])]\n",
    "\n",
    "# Create label vector\n",
    "Y = pandas.concat([merged_mirna_tumor, merged_mirna_normal])['pathologic_stage']\n",
    "Y.replace({'pathologic_stage': {'Stage I': 'tumor', 'Stage II': 'tumor', 'Stage III': 'tumor', 'Stage IV': 'tumor'}}, inplace=True)\n",
    "n_tumor = merged_mirna_tumor.shape[0]\n",
    "n_normal = merged_mirna_normal.shape[0]\n",
    "\n",
    "# Drop categorical columns\n",
    "merged_mirna_tumor.drop(['patient_barcode', 'pathologic_stage'], 1, inplace=True)\n",
    "merged_mirna_normal.drop(['patient_barcode', 'pathologic_stage'], 1, inplace=True)\n",
    "merged_gene_exp_tumor.drop(['patient_barcode'], 1, inplace=True)\n",
    "merged_gene_exp_normal.drop(['patient_barcode'], 1, inplace=True)\n",
    "\n",
    "# Drop duplicate columns names (Gene symbols with same name)\n",
    "_, i = np.unique(merged_gene_exp_tumor.columns, return_index=True)\n",
    "merged_gene_exp_tumor = merged_gene_exp_tumor.iloc[:, i]\n",
    "merged_gene_exp_normal = merged_gene_exp_normal.iloc[:, i]\n",
    "\n",
    "print \"merged_mirna_tumor\", merged_mirna_tumor.shape, 'nulls:', merged_mirna_tumor.isnull().sum().sum()\n",
    "print \"merged_mirna_normal\", merged_mirna_normal.shape, 'nulls:', merged_mirna_normal.isnull().sum().sum()\n",
    "print \"merged_gene_exp_tumor\", merged_gene_exp_tumor.shape, 'nulls:', merged_gene_exp_tumor.isnull().sum().sum()\n",
    "print \"merged_gene_exp_normal\", merged_gene_exp_normal.shape, 'nulls:', merged_gene_exp_normal.isnull().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Univariate feature selection via FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: TCGA-91-6849",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mValueError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-150-27301e62954b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mgene_exp_filter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSelectFdr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf_classif\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malpha\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.01\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mgen_exp_new\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_exp_filter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpandas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene_exp_tumor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene_exp_normal\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mY\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mprint\u001b[0m \u001b[0mmirna_new\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda/lib/python2.7/site-packages/sklearn/base.pyc\u001b[0m in \u001b[0;36mfit_transform\u001b[0;34m(self, X, y, **fit_params)\u001b[0m\n\u001b[1;32m    495\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    496\u001b[0m             \u001b[0;31m# fit method of arity 2 (supervised transformation)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 497\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    498\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda/lib/python2.7/site-packages/sklearn/feature_selection/univariate_selection.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y)\u001b[0m\n\u001b[1;32m    320\u001b[0m             \u001b[0mReturns\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    321\u001b[0m         \"\"\"\n\u001b[0;32m--> 322\u001b[0;31m         \u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_X_y\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'csr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'csc'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscore_func\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda/lib/python2.7/site-packages/sklearn/utils/validation.pyc\u001b[0m in \u001b[0;36mcheck_X_y\u001b[0;34m(X, y, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, warn_on_dtype, estimator)\u001b[0m\n\u001b[1;32m    519\u001b[0m     X = check_array(X, accept_sparse, dtype, order, copy, force_all_finite,\n\u001b[1;32m    520\u001b[0m                     \u001b[0mensure_2d\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_nd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensure_min_samples\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 521\u001b[0;31m                     ensure_min_features, warn_on_dtype, estimator)\n\u001b[0m\u001b[1;32m    522\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mmulti_output\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    523\u001b[0m         y = check_array(y, 'csr', force_all_finite=True, ensure_2d=False,\n",
      "\u001b[0;32m/opt/anaconda/lib/python2.7/site-packages/sklearn/utils/validation.pyc\u001b[0m in \u001b[0;36mcheck_array\u001b[0;34m(array, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)\u001b[0m\n\u001b[1;32m    400\u001b[0m         \u001b[0;31m# make sure we actually converted to numeric:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    401\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdtype_numeric\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkind\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"O\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 402\u001b[0;31m             \u001b[0marray\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfloat64\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    403\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mallow_nd\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m             raise ValueError(\"Found array with dim %d. %s expected <= 2.\"\n",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: TCGA-91-6849"
     ],
     "output_type": "error"
    }
   ],
   "source": [
    "from sklearn.feature_selection import SelectFdr, f_classif\n",
    "\n",
    "# Select the p-values for an estimated false discovery rate\n",
    "# FDR \n",
    "# mirna_filter = SelectFdr(f_classif, alpha=0.01)\n",
    "# mirna_new = mirna_filter.fit_transform(X=pandas.concat([mirna_tumor, mirna_normal]), y=Y)\n",
    "\n",
    "gene_exp_filter = SelectFdr(f_classif, alpha=0.01)\n",
    "gen_exp_new = gene_exp_filter.fit_transform(X=pandas.concat([gene_exp_tumor, gene_exp_normal]), y=Y)\n",
    "\n",
    "print mirna_new.shape\n",
    "print gen_exp_new.shape\n",
    "# mirna_tumor = mirna_tumor[mirna_tumor.columns[mirna_filter.get_support(indices=True)]]\n",
    "# mirna_normal = mirna_normal[mirna_normal.columns[mirna_filter.get_support(indices=True)]]\n",
    "gene_exp_tumor = gene_exp_tumor[gene_exp_tumor.columns[gene_exp_filter.get_support(indices=True)]]\n",
    "gene_exp_normal = gene_exp_normal[gene_exp_normal.columns[gene_exp_filter.get_support(indices=True)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mirna_tumor (506, 1883)\nmirna_normal (46, 1883)\ngene_tumor (515, 20486)\ngene_normal (59, 20486)\n"
     ]
    }
   ],
   "source": [
    "from src.data.make_dataset import TCGA_LUAD\n",
    "tgca_luad = TCGA_LUAD()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading TargetScan database for predicted miRNA-targets interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(185330, 2)\n"
     ]
    }
   ],
   "source": [
    "# Load data frame from file\n",
    "targetScan_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/external/TargetScan_Predicted_Targets_Info_default_predictions.tsv'), delimiter='\\t')\n",
    "targetScan_family_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/external/TargetScan_miR_Family_Info.txt'), delimiter='\\t')\n",
    "\n",
    "# Select only homo sapiens miRNA-target pairs\n",
    "targetScan_df = targetScan_df[targetScan_df[\"Species ID\"] == 9606][[\"miR Family\", \"Gene Symbol\"]]\n",
    "targetScan_family_df = targetScan_family_df[targetScan_family_df['Species ID'] == 9606][['miR family', 'MiRBase ID']]\n",
    "\n",
    "# Use miRBase ID names\n",
    "targetScan_family_df.rename(columns={'miR family': 'miR Family'}, inplace=True)\n",
    "targetScan_df = pandas.merge(targetScan_df, targetScan_family_df, how='inner', on=\"miR Family\")\n",
    "targetScan_df = targetScan_df[[\"MiRBase ID\", \"Gene Symbol\"]]\n",
    "\n",
    "# Standardize miRNA names\n",
    "targetScan_df['MiRBase ID'] = targetScan_df['MiRBase ID'].str.lower()\n",
    "targetScan_df['MiRBase ID'] = targetScan_df['MiRBase ID'].str.replace(\"-3p.*|-5p.*\", \"\")\n",
    "targetScan_df.drop_duplicates(inplace=True)\n",
    "\n",
    "# Filter miRNA-target pairs to only miRNA's included in miRNA expression data, same for gene targets\n",
    "targetScan_df = targetScan_df[targetScan_df['MiRBase ID'].isin(mirna_list) & targetScan_df['Gene Symbol'].isin(gene_symbols)]\n",
    "print targetScan_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build miRNA-target relationship network to from putative predicted miRNA-target pairs\n",
    "## Assign miRNA's to groups from the miRNA-target bipartite graph obtained above\n",
    "The assumption used is that two miRNA's are in the same group, if they share the same target\n",
    "If a miRNA has overlapping groups, \"larger_groups_first\" param decides whether larger or smaller group takes priority"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_A 20\nn_B 275\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6031\n"
     ]
    }
   ],
   "source": [
    "from src.models.miRNA_target_network import miRNATargetNetwork as miRNATargetNet\n",
    "import networkx as nx\n",
    "\n",
    "network = miRNATargetNet(miRNAs=merged_mirna_tumor.columns, targets=merged_gene_exp_tumor.columns, dys_threshold=0.6)\n",
    "\n",
    "print network.fit(tag=\"normal-StgI\",\n",
    "            miRNA_A=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='normal',\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_A=tgca_luad.make_dataset(dataset='gene', normal_tumor='normal',\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            miRNA_B=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage I'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_B=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage I'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            putative_assocs=targetScan_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_A 275\nn_B 120\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network.fit(tag=\"StgI-StgII\",\n",
    "            miRNA_A=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage I'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_A=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage I'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            miRNA_B=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage II'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_B=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage II'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            putative_assocs=targetScan_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_A 120\nn_B 84\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "397"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network.fit(tag=\"StgII-StgIII\",\n",
    "            miRNA_A=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage II'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_A=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage II'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            miRNA_B=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage III'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_B=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage III'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            putative_assocs=targetScan_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_A 84\nn_B 24\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5589"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "network.fit(tag=\"StgIII-StgIV\",\n",
    "            miRNA_A=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage III'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_A=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage III'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            miRNA_B=tgca_luad.make_dataset(dataset='miRNA', normal_tumor='tumor', pathologic_stages=['Stage IV'],\n",
    "                                           normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            gene_B=tgca_luad.make_dataset(dataset='gene', normal_tumor='tumor', pathologic_stages=['Stage IV'],\n",
    "                                          normal_matched=False, mirna_gene_matched=True)[0],\n",
    "            putative_assocs=targetScan_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1881\n20484\nedges 11656\n"
     ]
    }
   ],
   "source": [
    "# miRNAs_nodes, targets_nodes = nx.bipartite.sets(network.B)\n",
    "miRNAs_nodes = set(n for n, d in network.B.nodes(data=True) if d['bipartite']==0)\n",
    "targets_nodes = set(network.B) - miRNAs_nodes\n",
    "print miRNAs_nodes.__len__()\n",
    "print targets_nodes.__len__()\n",
    "\n",
    "edges = network.B.edges()\n",
    "miRNAs_nodes_degrees = nx.bipartite.degrees(network.B, miRNAs_nodes)[1]\n",
    "targets_nodes_degrees = nx.bipartite.degrees(network.B, targets_nodes)[1]\n",
    "\n",
    "print \"edges\", len(edges)\n",
    "# print \"miRNAs_nodes_degrees\", miRNAs_nodes_degrees\n",
    "# print \"targets_nodes_degrees\", targets_nodes_degrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare samples for Classification task"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X (552, 1881)\ny (552,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       " 1    506\n-1     46\nName: pathologic_stage, dtype: int64"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathologic_stage_map = {'normal': -1, 'Stage I': 1, 'Stage II': 1, 'Stage III': 1, 'Stage IV': 1}\n",
    "\n",
    "X, y = tgca_luad.make_dataset(dataset='miRNA', normal_tumor='both', normal_matched=False, mirna_gene_matched=False,\n",
    "                              pathologic_stages=['normal', 'Stage I', 'Stage II', 'Stage III', 'Stage IV'], label_mapping=pathologic_stage_map)\n",
    "\n",
    "print \"X\", X.shape\n",
    "print \"y\", y.shape\n",
    "\n",
    "y.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Create Classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X_train (441, 1881)\nY_train (441, 1) [-1.  1.]\nFISTA Flat: {'regul': 'group-lasso-l2', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FISTA Flat: {'regul': 'group-lasso-l2', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FISTA Flat: {'regul': 'l1', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}\nFISTA Flat: {'regul': 'elastic-net', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}\nFISTA Flat: {'regul': 'sparse-group-lasso-l2', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}\nFISTA Flat: {'regul': 'sparse-group-lasso-l2', 'verbose': True, 'numThreads': -1, 'max_it': 1000, 'pos': False, 'intercept': False, 'L0': 0.1, 'lambda1': 0.5, 'lambda2': 0.5, 'lambda3': 0.5, 'loss': 'weighted-logistic', 'it0': 10, 'tol': 0.001}"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[  1.46945370e-02],\n       [ -1.00000000e+00],\n       [ -1.00000000e+00],\n       [  6.77000000e+02]])"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from src.models.group_lasso_spams import SPAMSClassifier\n",
    "import spams\n",
    "\n",
    "myfloat = np.float\n",
    "param = {'numThreads': -1, 'verbose': True,\n",
    "         'it0': 10, 'max_it': 1000,\n",
    "         'L0': 0.1, 'tol': 1e-3, 'intercept': False,\n",
    "         'pos': False}\n",
    "param['lambda1'] = 0.5\n",
    "param['lambda2'] = 0.5\n",
    "param['lambda3'] = 0.5\n",
    "\n",
    "X_train, X_test, Y_train, Y_test = \\\n",
    "    train_test_split(X, y, test_size=0.2, random_state=np.random.randint(0, 10000), stratify=y)\n",
    "\n",
    "X_train = np.asfortranarray(X_train, dtype=myfloat)\n",
    "Y_train = np.asfortranarray(Y_train, dtype=myfloat)\n",
    "Y_train = np.reshape(Y_train, [Y_train.shape[0], 1])\n",
    "\n",
    "print \"X_train\", X_train.shape\n",
    "print \"Y_train\", Y_train.shape, np.unique(Y_train)\n",
    "\n",
    "param['loss'] = 'weighted-logistic'\n",
    "param['regul'] = 'group-lasso-l2'\n",
    "group_l2 = SPAMSClassifier()\n",
    "group_l2.fit(X_train, Y_train, groups=network.get_miRNA_group_assgn(mirna_list, smaller_groups=True), params=param)\n",
    "\n",
    "group_l2_large_groups = SPAMSClassifier()\n",
    "group_l2_large_groups.fit(X_train, Y_train, \n",
    "                          groups=network.get_miRNA_group_assgn(mirna_list, smaller_groups=False),\n",
    "                          params=param)\n",
    "\n",
    "param['regul'] = 'l1'\n",
    "l1_lasso = SPAMSClassifier()\n",
    "l1_lasso.fit(X_train, Y_train, groups=None, params=param)\n",
    "\n",
    "param['regul'] = 'elastic-net'\n",
    "elastic_net = SPAMSClassifier()\n",
    "elastic_net.fit(X_train, Y_train, groups=None, params=param)\n",
    "\n",
    "param['regul'] = 'sparse-group-lasso-l2'\n",
    "sparse_group_l2 = SPAMSClassifier()\n",
    "sparse_group_l2.fit(X_train, Y_train, groups=network.get_miRNA_group_assgn(mirna_list), params=param)\n",
    "\n",
    "sparse_group_l2_large_groups = SPAMSClassifier()\n",
    "sparse_group_l2_large_groups.fit(X_train, Y_train, groups=network.get_miRNA_group_assgn(mirna_list, smaller_groups=False), params=param)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression with L1L2 & Overlapping Group LASSO regularizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non-zero: 433\n"
     ]
    }
   ],
   "source": [
    "import parsimony.estimators as estimators\n",
    "import parsimony.algorithms as algorithms\n",
    "import parsimony.functions.nesterov.gl as gl\n",
    "import parsimony.utils.start_vectors as start_vectors\n",
    "# \n",
    "# X_train, X_test, Y_train, Y_test = \\\n",
    "#     train_test_split(X, y, test_size=0.2, random_state=np.random.randint(0, 10000), stratify=y)\n",
    "\n",
    "A = gl.linear_operator_from_groups(X_train.shape[1],\n",
    "                                   network.miRNA_groups_int\n",
    "                                   # weights=[2*1./len(g) for g in network.miRNA_groups_int]\n",
    "                                   )\n",
    "beta = start_vectors.ZerosStartVector().get_vector(X_train.shape[1])\n",
    "\n",
    "logistic_L1L2GL = estimators.LogisticRegressionL1L2GL(l1=2.5, l2=0.0, gl=3., A=A,\n",
    "                                                      algorithm=algorithms.proximal.FISTA(),\n",
    "                                                      class_weight='auto',\n",
    "                                                      algorithm_params=dict(max_iter=1000),\n",
    "                                                      mean=False)\n",
    "logistic_L1L2GL.fit(X=X_train, y=Y_train, beta=beta)\n",
    "\n",
    "print 'Non-zero:', len(logistic_L1L2GL.beta.nonzero()[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(array([ 1.]), 0.99999999999999978),\n (array([ 1.]), 1.0),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999999999827804),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999456252492824),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999999999999933),\n (array([ 0.]), 0.21821529936531853),\n (array([ 1.]), 0.99238325189969812),\n (array([ 0.]), 0.00084192209448990618),\n (array([ 1.]), 0.99997632858942287),\n (array([ 1.]), 0.99991994264051509),\n (array([ 1.]), 0.99996693084328669),\n (array([ 1.]), 0.99999823088406037),\n (array([ 1.]), 0.99999997828453413),\n (array([ 1.]), 0.99999972634093626),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.9999999999576481),\n (array([ 0.]), 6.0169912535891043e-06),\n (array([ 1.]), 0.99998360611126336),\n (array([ 1.]), 0.99973527443400856),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999999872392897),\n (array([ 1.]), 0.99999999307151444),\n (array([ 1.]), 0.99999999998453171),\n (array([ 1.]), 0.9999974256237959),\n (array([ 0.]), 2.9406639207016951e-07),\n (array([ 1.]), 0.96564533626123383),\n (array([ 1.]), 0.99948841686806178),\n (array([ 1.]), 0.99999999999999889),\n (array([ 0.]), 7.2194325554278297e-07),\n (array([ 1.]), 0.99999991962979951),\n (array([ 1.]), 0.68744101622141962),\n (array([ 1.]), 0.99999999999999356),\n (array([ 1.]), 0.9999994682260106),\n (array([ 1.]), 0.99999999999900036),\n (array([ 1.]), 0.69877017945912656),\n (array([ 1.]), 0.99999998740749396),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.9999998920581189),\n (array([ 1.]), 0.99999999852406396),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.91454265066429197),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999996989634377),\n (array([ 1.]), 0.99999999999965783),\n (array([ 1.]), 0.9999967086101933),\n (array([ 1.]), 0.99999999996351607),\n (array([ 1.]), 0.74250886959693141),\n (array([ 1.]), 0.99999999999999978),\n (array([ 1.]), 0.99999951226168304),\n (array([ 1.]), 0.9999999998758764),\n (array([ 1.]), 0.99999569234459773),\n (array([ 1.]), 0.99999999999999933),\n (array([ 1.]), 0.99999998888186592),\n (array([ 1.]), 0.99999999757089242),\n (array([ 1.]), 0.99999999999030398),\n (array([ 1.]), 0.99999999999999978),\n (array([ 0.]), 7.5781074881071611e-05),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999999999562794),\n (array([ 1.]), 0.99979877492485725),\n (array([ 1.]), 0.99999994422070038),\n (array([ 0.]), 6.1128535250401267e-05),\n (array([ 1.]), 0.99989906095789272),\n (array([ 1.]), 0.99999279135949459),\n (array([ 1.]), 0.99999999999466116),\n (array([ 1.]), 0.99999917785294867),\n (array([ 1.]), 0.99832600576214592),\n (array([ 1.]), 0.99996357699497995),\n (array([ 0.]), 9.7742281598512101e-05),\n (array([ 1.]), 0.99999999999987987),\n (array([ 1.]), 0.99989346803855039),\n (array([ 1.]), 0.99999999999458766),\n (array([ 1.]), 0.99999999993103783),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999999999839928),\n (array([ 1.]), 0.99999985975362382),\n (array([ 1.]), 0.99967260635442545),\n (array([ 1.]), 0.99807497775826981),\n (array([ 1.]), 0.99994125781255916),\n (array([ 1.]), 0.99999999736328493),\n (array([ 1.]), 0.9999999839109186),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99999498263041153),\n (array([ 1.]), 0.99999999999933564),\n (array([ 1.]), 0.99999999437588194),\n (array([ 1.]), 0.99999798233281889),\n (array([ 1.]), 0.99999632392097837),\n (array([ 0.]), 0.18670609824181858),\n (array([ 1.]), 0.9999999999643161),\n (array([ 1.]), 0.99994162919903062),\n (array([ 0.]), 0.0040217234524108575),\n (array([ 1.]), 0.99999990579765252),\n (array([ 1.]), 0.99999999999999867),\n (array([ 1.]), 0.99998199981193325),\n (array([ 1.]), 0.99999999996162448),\n (array([ 1.]), 0.99998388598541377),\n (array([ 1.]), 0.9999999999999909),\n (array([ 1.]), 0.99998097552818588),\n (array([ 1.]), 0.9999999999999678),\n (array([ 1.]), 0.99999999999995715),\n (array([ 1.]), 0.99999999930392125),\n (array([ 1.]), 1.0),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99997759788758289),\n (array([ 1.]), 0.9999999998500777),\n (array([ 1.]), 1.0),\n (array([ 1.]), 1.0),\n (array([ 1.]), 0.99964742150682506)]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(label, percent[0]) for label, percent in zip(logistic_L1L2GL.predict(X_test), logistic_L1L2GL.predict_probability(X_test))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Classification Report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\nLogistic L1L2GL\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\ngroup_small\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\ngroup_large\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\nl1_lasso\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\nelastic_net\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\nsparse-group_small\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n\nsparse_group_large\n             precision    recall  f1-score   support\n\n       -1.0       0.97      1.00      0.99        37\n        1.0       1.00      1.00      1.00       404\n\navg / total       1.00      1.00      1.00       441\n\n             precision    recall  f1-score   support\n\n         -1       0.90      1.00      0.95         9\n          1       1.00      0.99      1.00       102\n\navg / total       0.99      0.99      0.99       111\n\n"
     ]
    }
   ],
   "source": [
    "from sklearn import metrics\n",
    "\n",
    "for model, name in [(logistic_L1L2GL, 'Logistic L1L2GL'),\n",
    "                    (group_l2, 'group_small'),\n",
    "                    (group_l2_large_groups, 'group_large'),\n",
    "                    (l1_lasso, 'l1_lasso'),\n",
    "                    (elastic_net, 'elastic_net'),\n",
    "                    (sparse_group_l2, 'sparse-group_small'),\n",
    "                    (sparse_group_l2_large_groups, 'sparse_group_large')]:\n",
    "    print '\\n', name\n",
    "    if name is 'Logistic L1L2GL':\n",
    "        predicted = logistic_L1L2GL.predict(X_train)\n",
    "        np.place(predicted, predicted == 0., [-1])\n",
    "        print metrics.classification_report(Y_train, predicted)\n",
    "        predicted = logistic_L1L2GL.predict(X_test)\n",
    "        np.place(predicted, predicted == 0., [-1])\n",
    "        print metrics.classification_report(Y_test, predicted)\n",
    "    else:\n",
    "        print metrics.classification_report(Y_train, model.predict(X_train))\n",
    "\n",
    "        print metrics.classification_report(Y_test, model.predict(X_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading HMDD database for experimentally validated NSCLC miRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMDD_lung_neoplasm_miRNAs (157,)\nmiRCancer_luad_miRNAs 27\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mir2disease_nsclc_miRNAs (24,)\n"
     ]
    }
   ],
   "source": [
    "HMDD_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/external/HMDD_miRNA-disease.txt'), delimiter='\\t', header=-1)\n",
    "HMDD_df = HMDD_df[HMDD_df[2] == \"Lung Neoplasms\"]\n",
    "HMDD_lung_neoplasm_miRNAs = HMDD_df[1].unique()\n",
    "print \"HMDD_lung_neoplasm_miRNAs\", HMDD_lung_neoplasm_miRNAs.shape\n",
    "\n",
    "miRCancer_luad_miRNA_df = pandas.read_csv(os.path.join(ROOT_DIR, 'data/external/validated_luad_miRNAs_miRCancer.csv'))\n",
    "miRCancer_luad_miRNAs = set(miRCancer_luad_miRNA_df['miRNA'].tolist())\n",
    "print \"miRCancer_luad_miRNAs\", len(miRCancer_luad_miRNAs)\n",
    "\n",
    "\n",
    "mir2disease_nsclc_miRNA_df = pandas.read_table(os.path.join(ROOT_DIR, 'data/external/mir2Disease_miRNA-Disease.txt'), delimiter='\\t', header=-1) \n",
    "mir2disease_nsclc_miRNA_df.columns = ['miRNA', 'disease', 'up_down', 'method', 'year', 'description'] \n",
    "mir2disease_nsclc_miRNA_df = mir2disease_nsclc_miRNA_df[(mir2disease_nsclc_miRNA_df['disease'] == 'non-small cell lung cancer (NSCLC)')]\n",
    "mir2disease_nsclc_miRNA_df['miRNA'] = mir2disease_nsclc_miRNA_df['miRNA'].str.lower()\n",
    "mir2disease_nsclc_miRNA_df['miRNA'] = mir2disease_nsclc_miRNA_df['miRNA'].str.replace(\"*\", \"\")\n",
    "mir2disease_nsclc_miRNA_df['miRNA'] = mir2disease_nsclc_miRNA_df['miRNA'].str.replace(\"-3p.*|-5p.*\", \"\")\n",
    "mir2disease_nsclc_miRNA_df.drop_duplicates(inplace=True)\n",
    "mir2disease_nsclc_miRNAs = mir2disease_nsclc_miRNA_df['miRNA'].unique()\n",
    "print \"mir2disease_nsclc_miRNAs\", mir2disease_nsclc_miRNAs.shape\n",
    "\n",
    "valudating_miRNAs = HMDD_lung_neoplasm_miRNAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare models by recall rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>method</th>\n",
       "      <th>no_selected</th>\n",
       "      <th>recall_rate</th>\n",
       "      <th>precision_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Logistic L1L2GL</td>\n",
       "      <td>433</td>\n",
       "      <td>0.847134</td>\n",
       "      <td>0.307159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>group_large</td>\n",
       "      <td>252</td>\n",
       "      <td>0.707006</td>\n",
       "      <td>0.440476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>group_small</td>\n",
       "      <td>174</td>\n",
       "      <td>0.630573</td>\n",
       "      <td>0.568966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>l1_lasso</td>\n",
       "      <td>150</td>\n",
       "      <td>0.592357</td>\n",
       "      <td>0.620000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>elastic_net</td>\n",
       "      <td>150</td>\n",
       "      <td>0.592357</td>\n",
       "      <td>0.620000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sparse_group_large</td>\n",
       "      <td>130</td>\n",
       "      <td>0.515924</td>\n",
       "      <td>0.623077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sparse-group_small</td>\n",
       "      <td>128</td>\n",
       "      <td>0.509554</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>method</th>\n",
       "      <th>no_selected</th>\n",
       "      <th>recall_rate</th>\n",
       "      <th>precision_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Logistic L1L2GL</td>\n",
       "      <td>433</td>\n",
       "      <td>0.847134</td>\n",
       "      <td>0.307159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>group_large</td>\n",
       "      <td>252</td>\n",
       "      <td>0.707006</td>\n",
       "      <td>0.440476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>group_small</td>\n",
       "      <td>174</td>\n",
       "      <td>0.630573</td>\n",
       "      <td>0.568966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>l1_lasso</td>\n",
       "      <td>150</td>\n",
       "      <td>0.592357</td>\n",
       "      <td>0.620000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>elastic_net</td>\n",
       "      <td>150</td>\n",
       "      <td>0.592357</td>\n",
       "      <td>0.620000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sparse_group_large</td>\n",
       "      <td>130</td>\n",
       "      <td>0.515924</td>\n",
       "      <td>0.623077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sparse-group_small</td>\n",
       "      <td>128</td>\n",
       "      <td>0.509554</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.utils.validate_candidate_miRNAs import recall_rate, precision_rate, get_miRNA_names\n",
    "\n",
    "models = []\n",
    "for model, name in [(group_l2, 'group_small'),\n",
    "                    (logistic_L1L2GL, 'Logistic L1L2GL'),\n",
    "                    (group_l2_large_groups, 'group_large'),\n",
    "                    (l1_lasso, 'l1_lasso'),\n",
    "                    (elastic_net, 'elastic_net'),\n",
    "                    (sparse_group_l2, 'sparse-group_small'),\n",
    "                    (sparse_group_l2_large_groups, 'sparse_group_large')]:\n",
    "    if name is 'Logistic L1L2GL':\n",
    "        models.append([name,\n",
    "                   len(model.beta.nonzero()[0]),\n",
    "                   recall_rate(candidate_miRNAs=get_miRNA_names(model.beta.nonzero()[0][:], mirna_list),\n",
    "                               validated_miRNAs=valudating_miRNAs),\n",
    "                   precision_rate(candidate_miRNAs=get_miRNA_names(model.beta.nonzero()[0][:], mirna_list),\n",
    "                                  validated_miRNAs=valudating_miRNAs) ])\n",
    "    else:\n",
    "        models.append([name,\n",
    "                   model.get_selected_miRNAs(mirna_list)['miRNA'].shape[0],\n",
    "                   recall_rate(candidate_miRNAs=model.get_selected_miRNAs(mirna_list)['miRNA'].tolist(),\n",
    "                               validated_miRNAs=valudating_miRNAs),\n",
    "                   precision_rate(candidate_miRNAs=model.get_selected_miRNAs(mirna_list)['miRNA'].tolist(),\n",
    "                                  validated_miRNAs=valudating_miRNAs) ])\n",
    "\n",
    "pandas.DataFrame(models, columns=['method', 'no_selected', 'recall_rate', 'precision_rate']).sort_values(by='recall_rate', ascending=False)\n",
    "# (group_l2_large_groups, 'group_l2_large_groups'),\n",
    "# (l1_lasso, 'l1_lasso'),\n",
    "# (elastic_net, 'elastic_net'),\n",
    "# (sparse_group_l2, 'sparse-group-lasso-l2'),\n",
    "# (sparse_group_l2_large_groups, 'sparse_group_l2_large_groups')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check each candidate miRNA's if validated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>coef</th>\n",
       "      <th>validated</th>\n",
       "      <th>groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>hsa-mir-203a</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>False</td>\n",
       "      <td>[23, 65, 343, 449, 651, 700, 864, 1082, 1164, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>hsa-mir-148a</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>True</td>\n",
       "      <td>[228, 242, 355, 641, 763, 793, 801, 888, 909, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>hsa-mir-182</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>True</td>\n",
       "      <td>[34, 190, 254, 301, 335, 455, 509, 600, 715, 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>True</td>\n",
       "      <td>[157, 235, 457, 608, 655, 804, 1016, 1140, 129...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hsa-mir-103a-2</td>\n",
       "      <td>-0.000049</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hsa-mir-103a-1</td>\n",
       "      <td>-0.000048</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>hsa-mir-29c</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[276, 335, 339, 369, 469, 484, 563, 582, 588, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>hsa-mir-99b</td>\n",
       "      <td>-0.000044</td>\n",
       "      <td>False</td>\n",
       "      <td>[1247, 1255, 1348]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>hsa-mir-30e</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>True</td>\n",
       "      <td>[257, 614, 904, 966, 1001, 1035, 1296, 1323, 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>hsa-mir-183</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>True</td>\n",
       "      <td>[118, 182, 226, 231, 246, 291, 318, 419, 423, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>hsa-mir-30a</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>True</td>\n",
       "      <td>[263, 315, 353, 363, 486, 487, 520, 551, 554, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>hsa-mir-143</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[146, 168, 255, 277, 354, 359, 734, 736, 816, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hsa-let-7f-1</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>-0.000025</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>-0.000025</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>hsa-mir-92a-1</td>\n",
       "      <td>-0.000023</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>True</td>\n",
       "      <td>[17, 45, 69, 152, 196, 336, 365, 407, 468, 472...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>hsa-mir-92a-2</td>\n",
       "      <td>-0.000022</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>-0.000021</td>\n",
       "      <td>True</td>\n",
       "      <td>[422, 517, 681, 702, 823, 982, 1000, 1022, 108...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>coef</th>\n",
       "      <th>validated</th>\n",
       "      <th>groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>hsa-mir-203a</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>False</td>\n",
       "      <td>[23, 65, 343, 449, 651, 700, 864, 1082, 1164, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>hsa-mir-148a</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>True</td>\n",
       "      <td>[228, 242, 355, 641, 763, 793, 801, 888, 909, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>hsa-mir-182</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>True</td>\n",
       "      <td>[34, 190, 254, 301, 335, 455, 509, 600, 715, 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>True</td>\n",
       "      <td>[157, 235, 457, 608, 655, 804, 1016, 1140, 129...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hsa-mir-103a-2</td>\n",
       "      <td>-0.000049</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hsa-mir-103a-1</td>\n",
       "      <td>-0.000048</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>hsa-mir-29c</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[276, 335, 339, 369, 469, 484, 563, 582, 588, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>hsa-mir-99b</td>\n",
       "      <td>-0.000044</td>\n",
       "      <td>False</td>\n",
       "      <td>[1247, 1255, 1348]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>hsa-mir-30e</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>True</td>\n",
       "      <td>[257, 614, 904, 966, 1001, 1035, 1296, 1323, 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>hsa-mir-183</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>True</td>\n",
       "      <td>[118, 182, 226, 231, 246, 291, 318, 419, 423, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>hsa-mir-30a</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>True</td>\n",
       "      <td>[263, 315, 353, 363, 486, 487, 520, 551, 554, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>hsa-mir-143</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[146, 168, 255, 277, 354, 359, 734, 736, 816, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hsa-let-7f-1</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>-0.000025</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>-0.000025</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>hsa-mir-92a-1</td>\n",
       "      <td>-0.000023</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>True</td>\n",
       "      <td>[17, 45, 69, 152, 196, 336, 365, 407, 468, 472...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>hsa-mir-92a-2</td>\n",
       "      <td>-0.000022</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>-0.000021</td>\n",
       "      <td>True</td>\n",
       "      <td>[422, 517, 681, 702, 823, 982, 1000, 1022, 108...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_miRNAs = group_l2_large_groups.get_selected_miRNAs(mirna_list)\n",
    "candidate_miRNAs['validated'] = candidate_miRNAs.apply(lambda row: row['miRNA'] in HMDD_lung_neoplasm_miRNAs, axis=1)\n",
    "candidate_miRNAs['groups'] = candidate_miRNAs.apply(lambda row: network.find_miRNA_groups(row['miRNA']), axis=1)\n",
    "candidate_miRNAs.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:5: FutureWarning: sort is deprecated, use sort_values(inplace=True) for INPLACE sorting\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>coef</th>\n",
       "      <th>validated</th>\n",
       "      <th>groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>hsa-mir-203a</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>False</td>\n",
       "      <td>[23, 65, 343, 449, 651, 700, 864, 1082, 1164, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>hsa-mir-148a</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>True</td>\n",
       "      <td>[228, 242, 355, 641, 763, 793, 801, 888, 909, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>hsa-mir-182</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>True</td>\n",
       "      <td>[34, 190, 254, 301, 335, 455, 509, 600, 715, 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hsa-mir-103a-2</td>\n",
       "      <td>-0.000070</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hsa-mir-103a-1</td>\n",
       "      <td>-0.000069</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>True</td>\n",
       "      <td>[157, 235, 457, 608, 655, 804, 1016, 1140, 129...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>hsa-mir-29c</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>True</td>\n",
       "      <td>[276, 335, 339, 369, 469, 484, 563, 582, 588, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>hsa-mir-99b</td>\n",
       "      <td>-0.000056</td>\n",
       "      <td>False</td>\n",
       "      <td>[1247, 1255, 1348]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>hsa-mir-30e</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>True</td>\n",
       "      <td>[257, 614, 904, 966, 1001, 1035, 1296, 1323, 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>-0.000046</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hsa-let-7f-1</td>\n",
       "      <td>-0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>hsa-mir-183</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[118, 182, 226, 231, 246, 291, 318, 419, 423, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>hsa-mir-30a</td>\n",
       "      <td>-0.000038</td>\n",
       "      <td>True</td>\n",
       "      <td>[263, 315, 353, 363, 486, 487, 520, 551, 554, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>hsa-mir-143</td>\n",
       "      <td>-0.000038</td>\n",
       "      <td>True</td>\n",
       "      <td>[146, 168, 255, 277, 354, 359, 734, 736, 816, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>-0.000035</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>-0.000034</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>hsa-mir-92a-1</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>hsa-mir-92a-2</td>\n",
       "      <td>-0.000030</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>hsa-mir-451a</td>\n",
       "      <td>-0.000029</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>True</td>\n",
       "      <td>[422, 517, 681, 702, 823, 982, 1000, 1022, 108...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>coef</th>\n",
       "      <th>validated</th>\n",
       "      <th>groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>hsa-mir-203a</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>False</td>\n",
       "      <td>[23, 65, 343, 449, 651, 700, 864, 1082, 1164, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>hsa-mir-148a</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>True</td>\n",
       "      <td>[228, 242, 355, 641, 763, 793, 801, 888, 909, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>hsa-mir-182</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>True</td>\n",
       "      <td>[34, 190, 254, 301, 335, 455, 509, 600, 715, 7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hsa-mir-103a-2</td>\n",
       "      <td>-0.000070</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hsa-mir-103a-1</td>\n",
       "      <td>-0.000069</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>True</td>\n",
       "      <td>[157, 235, 457, 608, 655, 804, 1016, 1140, 129...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>hsa-mir-29c</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>True</td>\n",
       "      <td>[276, 335, 339, 369, 469, 484, 563, 582, 588, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>hsa-mir-99b</td>\n",
       "      <td>-0.000056</td>\n",
       "      <td>False</td>\n",
       "      <td>[1247, 1255, 1348]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>hsa-mir-30e</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>True</td>\n",
       "      <td>[257, 614, 904, 966, 1001, 1035, 1296, 1323, 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hsa-let-7f-2</td>\n",
       "      <td>-0.000046</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hsa-let-7f-1</td>\n",
       "      <td>-0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>hsa-mir-183</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>True</td>\n",
       "      <td>[118, 182, 226, 231, 246, 291, 318, 419, 423, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>hsa-mir-30a</td>\n",
       "      <td>-0.000038</td>\n",
       "      <td>True</td>\n",
       "      <td>[263, 315, 353, 363, 486, 487, 520, 551, 554, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>hsa-mir-143</td>\n",
       "      <td>-0.000038</td>\n",
       "      <td>True</td>\n",
       "      <td>[146, 168, 255, 277, 354, 359, 734, 736, 816, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>-0.000035</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>-0.000034</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>hsa-mir-92a-1</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>True</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>hsa-mir-92a-2</td>\n",
       "      <td>-0.000030</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>hsa-mir-451a</td>\n",
       "      <td>-0.000029</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>hsa-mir-22</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>True</td>\n",
       "      <td>[422, 517, 681, 702, 823, 982, 1000, 1022, 108...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_miRNAs = pandas.DataFrame([(mirna_list[m], logistic_L1L2GL.beta[m][0]) for m in logistic_L1L2GL.beta.nonzero()[0]])\n",
    "candidate_miRNAs.columns = ['miRNA', 'coef']\n",
    "candidate_miRNAs['validated'] = candidate_miRNAs.apply(lambda row: row['miRNA'] in HMDD_lung_neoplasm_miRNAs, axis=1)\n",
    "candidate_miRNAs['groups'] = candidate_miRNAs.apply(lambda row: network.find_miRNA_groups(row['miRNA']), axis=1)\n",
    "candidate_miRNAs.reindex(candidate_miRNAs.coef.abs().sort(inplace=False, ascending=False).index).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validate miRNA-target edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ITGA4 hsa-mir-301b\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'miRecords_df' is not defined",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mNameError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-54-73141c8de670>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         validated_edges = pandas.concat([validated_edges,\n\u001b[0;32m---> 10\u001b[0;31m                                          miRecords_df[(miRecords_df[\"miRNA_mature_ID\"] == v2) &\n\u001b[0m\u001b[1;32m     11\u001b[0m                                                       (miRecords_df[\"Target gene_name\"] == v1)]])\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'miRecords_df' is not defined"
     ],
     "output_type": "error"
    }
   ],
   "source": [
    "validated_edges = pandas.DataFrame()\n",
    "for (v1, v2) in edges:\n",
    "    print v1, v2\n",
    "    if network.B.node[v1]['bipartite'] == 0:  # if v1 is miRNA\n",
    "        validated_edges = pandas.concat([validated_edges,\n",
    "                                         miRecords_df[(miRecords_df[\"miRNA_mature_ID\"] == v1) &\n",
    "                                                      (miRecords_df[\"Target gene_name\"] == v2)]])\n",
    "    else:\n",
    "        validated_edges = pandas.concat([validated_edges,\n",
    "                                         miRecords_df[(miRecords_df[\"miRNA_mature_ID\"] == v2) &\n",
    "                                                      (miRecords_df[\"Target gene_name\"] == v1)]])\n",
    "\n",
    "print float(len(validated_edges))/float(len(edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY8AAAEKCAYAAADq59mMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGnRJREFUeJzt3XuUZWV95vHvI4gXlHBrHKRpG0xjREdb6EGMl0XUIKIR\nNF5giaKw0l4g0VEzQuIsHRNmNF5ImBgUFbmoIIpgqygg8RYHhG5AaERCCx1paekWFVEMy4bf/LF3\n0Yfuquqzizp1TlHfz1pn1dnv2ZdftUee2u+797tTVUiS1MVDhl2AJGn2MTwkSZ0ZHpKkzgwPSVJn\nhockqTPDQ5LUmeEhSerM8JAkdWZ4SJI623rYBQzKzjvvXAsXLhx2GZI0a6xYseLnVTWvn3UftOGx\ncOFCli9fPuwyJGnWSPIf/a5rt5UkqTPDQ5LUmeEhSerM8JAkdWZ4SJI6G1h4JDk1ybokK3vaPpfk\n6va1OsnVbfvCJL/r+eyjPdvsm+TaJKuSnJQkg6pZktSfQV6qexrwz8AZYw1V9aqx90k+BNzRs/6P\nq2rxOPs5GVgKXAZcABwEfG0A9UqS+jSwM4+q+g7wi/E+a88eXgmcNdk+kuwKbFdVl1bzvNwzgEOn\nu1ZJUjfDGvN4NnBbVd3Y07ZHkquSfDvJs9u23YA1PeusadskSUM0rDvMD+f+Zx1rgQVVdXuSfYHz\nkzwJGG98oybaaZKlNF1cLFiwYMrFLTzuq32tt/p9L5ryMSRpNpvxM48kWwMvAz431lZVd1fV7e37\nFcCPgb1ozjTm92w+H7h1on1X1SlVtaSqlsyb19f0LJKkKRhGt9XzgR9V1X3dUUnmJdmqfb8nsAi4\nqarWAncm2b8dJ3kt8KUh1CxJ6jHIS3XPAi4FnpBkTZKj248OY/OB8ucA1yT5AfAF4I1VNTbY/ibg\nE8AqmjMSr7SSpCEb2JhHVR0+Qfvrxmk7Fzh3gvWXA0+e1uIkSQ+Id5hLkjozPCRJnRkekqTODA9J\nUmeGhySpM8NDktSZ4SFJ6szwkCR1ZnhIkjozPCRJnRkekqTODA9JUmeGhySpM8NDktSZ4SFJ6szw\nkCR1ZnhIkjozPCRJnRkekqTODA9JUmeGhySps4GFR5JTk6xLsrKn7T1Jfprk6vZ1cM9nxydZleSG\nJC/oaT+obVuV5LhB1StJ6t8gzzxOAw4ap/3Eqlrcvi4ASLI3cBjwpHabf0myVZKtgI8ALwT2Bg5v\n15UkDdHWg9pxVX0nycI+Vz8EOLuq7gZuTrIK2K/9bFVV3QSQ5Ox23R9Oc7mSpA6GMeZxbJJr2m6t\nHdq23YBbetZZ07ZN1D6uJEuTLE+yfP369dNdtySpNdPhcTLweGAxsBb4UNuecdatSdrHVVWnVNWS\nqloyb968B1qrJGkCA+u2Gk9V3Tb2PsnHga+0i2uA3XtWnQ/c2r6fqF2SNCQzeuaRZNeexZcCY1di\nLQMOS/KwJHsAi4DLgSuARUn2SLINzaD6spmsWZK0uYGdeSQ5CzgA2DnJGuDdwAFJFtN0Pa0G3gBQ\nVdclOYdmIHwDcExV3dPu51jgQmAr4NSqum5QNUuS+jPIq60OH6f5k5OsfwJwwjjtFwAXTGNpkqQH\nyDvMJUmdGR6SpM4MD0lSZ4aHJKkzw0OS1JnhIUnqzPCQJHVmeEiSOjM8JEmdGR6SpM4MD0lSZ4aH\nJKkzw0OS1JnhIUnqzPCQJHVmeEiSOjM8JEmdGR6SpM4MD0lSZ4aHJKmzgYVHklOTrEuysqftA0l+\nlOSaJOcl2b5tX5jkd0mubl8f7dlm3yTXJlmV5KQkGVTNkqT+DPLM4zTgoE3aLgaeXFVPAf4dOL7n\nsx9X1eL29cae9pOBpcCi9rXpPiVJM2xg4VFV3wF+sUnbRVW1oV28DJg/2T6S7ApsV1WXVlUBZwCH\nDqJeSVL/hjnmcRTwtZ7lPZJcleTbSZ7dtu0GrOlZZ03bJkkaoq2HcdAkfwtsAD7TNq0FFlTV7Un2\nBc5P8iRgvPGNmmS/S2m6uFiwYMH0Fi1Jus8WzzySvCXJdml8MsmVSQ6c6gGTHAm8GHh12xVFVd1d\nVbe371cAPwb2ojnT6O3amg/cOtG+q+qUqlpSVUvmzZs31RIlSVvQT7fVUVX1a+BAYAfgNcD7pnKw\nJAcB7wReUlV39bTPS7JV+35PmoHxm6pqLXBnkv3bq6xeC3xpKseWJE2ffrqtxrqODgbOrKrr+rlc\nNslZwAHAzknWAO+mubrqYcDF7S4ua6+seg7w3iQbgHuAN1bV2GD7m2iu3HoEzRhJ7ziJJGkI+gmP\nFUkuAvYAjk/yaODeLW1UVYeP0/zJCdY9Fzh3gs+WA0/uo05J0gzpJzyOBhbTdCPdlWQn4PWDLUuS\nNMr6GfMoYG/gr9rlbYGHD6wiSdLI6yc8/gV4BjDWDXUn8JGBVSRJGnn9dFs9var2SXIVQFX9Msk2\nA65LkjTC+jnz+H17GW1Bc1ktfQyYS5IevPoJj5OA84BdkpwA/BvwvwdalSRppG2x26qqPpNkBfA8\nmns+Dq2q6wdemSRpZE0YHkl27FlcB5zV+1nPTXySpDlmsjOPFTTjHAEWAL9s328P/ITmpkFJ0hw0\n4ZhHVe1RVXsC3wD+rKp2rqqdaCY1vGimCpQkjZ5+Bsz3r6oLxhaq6mvAHw+uJEnSqOvnPo9bk7wL\n+DRNN9YRTDItuiTpwa+fM4/DgXk0l+ueD+zCxrvNJUlzUD+X6v4CeEuS7YB7q+o3gy9LkjTK+nmS\n4H9tpya5FrguyYokTpEuSXNYP91WHwPeVlWPq6rHAW8HThlsWZKkUdZPeGxbVd8cW6iqb9FMyy5J\nmqP6udrqpiT/EzizXT4CuHlwJUmSRl0/Zx5H0Vxt9UWaK67m4ZMEJWlO6+dqq1+y8SmCkiRtOTyS\n7AW8A1jYu35VPXdwZUmSRlk/3VafB64C3gX8dc9ri5KcmmRdkpU9bTsmuTjJje3PHdr2JDkpyaok\n1yTZp2ebI9v1b0xyZJdfUJI0/foJjw1VdXJVXV5VK8Zefe7/NOCgTdqOAy6pqkXAJe0ywAuBRe1r\nKXAy3Dc1/LuBpwP7Ae8eCxxJ0nD0Ex5fTvLmJLu2Zw07bvKsjwlV1XeATZ/7cQhwevv+dODQnvYz\nqnEZsH2SXYEXABdX1S/a8ZeL2TyQJEkzqJ9Ldce6iXq7qgrYc4rHfExVrQWoqrVJdmnbdwNu6Vlv\nTds2UftmkiylOWthwYIFUyxPkrQl/VxtNVMPfcp4h5+kffPGqlNo735fsmTJuOtIkh64frqtpttt\nbXcU7c91bfsaYPee9ebTTP0+UbskaUiGER7L2NgVdiTwpZ7217ZXXe0P3NF2b10IHJhkh3ag/MC2\nTZI0JBN2WyV5ZlV9L8nDquruqew8yVnAAcDOSdbQXDX1PuCcJEfTPAv9Fe3qFwAHA6uAu2jvYq+q\nXyT5O+CKdr33ttPES5KGZLIxj5OAfYFLgX0mWW9CVTXRQ6OeN866BRwzwX5OBU6dSg2SpOk3WXj8\nPskpwG5JTtr0w6pyyhJJmqMmC48XA8+nuc+i35sCJUlzwIThUVU/B85Ocn1V/WAGa5Ikjbh+rra6\nPcl57RxVtyU5N8n8gVcmSRpZ/YTHp2guo30szZ3dX27bJElzVD/hsUtVfaqqNrSv02geCCVJmqP6\nCY/1SY5IslX7OgK4fdCFSZJGV7+PoX0l8DNgLfDytk2SNEf1MzHiT4CXzEAtkqRZYhhzW0mSZjnD\nQ5LUmeEhSepsi+GR5C1JtmunSv9kkiuTHDgTxUmSRlNfV1tV1a9pnqOxA/AammnVJUlzVD/hMfYY\n2IOBM6vqOsZ/NKwkaY7oJzxWJLmIJjwuTPJo4N7BliVJGmVbvM8DOBpYDNxUVXcl2Yn2KX+SpLmp\nnzOPi6vqyqr6FUBV3Q6cONiyJEmjbLJnmD8ceCTN88d3YOM4x3Y0M+xKkuaoybqt3gC8lSYoVrAx\nPH4NfGTAdUmSRtiE3VZV9U9VtQfwjqras6r2aF9Prap/nuoBkzwhydU9r18neWuS9yT5aU/7wT3b\nHJ9kVZIbkrxgqseWJE2PfiZG/L9J/hhY2Lt+VZ0xlQNW1Q00A/Ak2Qr4KXAezSD8iVX1wd71k+wN\nHAY8ieYs6BtJ9qqqe6ZyfEnSA7fF8EhyJvB44Gpg7D/YBUwpPDbxPODHVfUfyYS3jhwCnF1VdwM3\nJ1kF7AdcOg3HlyRNQT+X6i4B9q6qGsDxDwPO6lk+NslrgeXA26vqlzSPvr2sZ501bZskaUj6uVR3\nJfBfpvvASbaheU7I59umk2nOcBbTPHTqQ2OrjrP5uEGWZGmS5UmWr1+/fporliSN6efMY2fgh0ku\nB+4ea6yqB/qAqBcCV1bVbe3+bhv7IMnHga+0i2uA3Xu2mw/cOt4Oq+oU4BSAJUuWDOJMSZJEf+Hx\nngEd+3B6uqyS7FpVa9vFl9Kc8QAsAz6b5MM0A+aLgMsHVJMkqQ/9XG317ek+aJJHAn9Kcy/JmH9I\nspimS2r12GdVdV2Sc4AfAhuAY7zSSpKGq5+rre5k4xjDNsBDgd9W1XZTPWhV3QXstEnbayZZ/wTg\nhKkeT5I0vfo583h073KSQ4H9B1aRJGnkdX4MbVWdD3iXtyTNYf10W72sZ/EhNPd9/OfAKpIkjbx+\nrrb6s573G2gGsw8ZSDWSpFmhnzEPH/wkSbqfLY55JJmf5Lwk65LcluTcJPNnojhJ0mjqZ8D8UzQ3\n6j2WZk6pL7dtkqQ5qp/wmFdVn6qqDe3rNGDegOuSJI2wfsLj50mOSLJV+zoCuH3QhUmSRlc/4XEU\n8ErgZzSz3b68bZMkzVH9XG31E5qp0yVJAvq72ur0JNv3LO+Q5NTBliVJGmX9dFs9pap+NbbQPt3v\naYMrSZI06voJj4ck2WFsIcmO9HdnuiTpQaqfEPgQ8P+SfIFmavZX4vTokjSn9TNgfkaS5cBzaZ4n\n/rKq+uHAK5Mkjay+up/asDAwJEnAFJ7nIUmS4SFJ6szwkCR1ZnhIkjobWngkWZ3k2iRXt1dzkWTH\nJBcnubH9uUPbniQnJVmV5Jok+wyrbknS8M88/qSqFlfVknb5OOCSqloEXNIuA7wQWNS+lgInz3il\nkqT7DDs8NnUIcHr7/nTg0J72M6pxGbB9kl2HUaAkabjhUcBFSVYkWdq2Paaq1gK0P3dp23cDbunZ\ndk3bJkkagmHOUfXMqro1yS7AxUl+NMm6GaetNlupCaGlAAsWLJieKiVJmxnamUdV3dr+XAecB+wH\n3DbWHdX+XNeuvgbYvWfz+cCt4+zzlKpaUlVL5s3zSbmSNChDCY8k2yZ59Nh74EBgJbAMOLJd7Ujg\nS+37ZcBr26uu9gfuGOvekiTNvGF1Wz0GOC/JWA2fraqvJ7kCOCfJ0cBPgFe0618AHAysAu4CXj/z\nJUuSxgwlPKrqJuCp47TfDjxvnPYCjpmB0iRJfRi1S3UlSbOA4SFJ6szwkCR1ZnhIkjozPCRJnRke\nkqTODA9JUmeGhySpM8NDktSZ4SFJ6szwkCR1ZnhIkjozPCRJnRkekqTODA9JUmeGhySpM8NDktSZ\n4SFJ6szwkCR1ZnhIkjozPCRJnc14eCTZPck3k1yf5Lokb2nb35Pkp0mubl8H92xzfJJVSW5I8oKZ\nrlmSdH9bD+GYG4C3V9WVSR4NrEhycfvZiVX1wd6Vk+wNHAY8CXgs8I0ke1XVPTNatSTpPjN+5lFV\na6vqyvb9ncD1wG6TbHIIcHZV3V1VNwOrgP0GX6kkaSJDHfNIshB4GvD9tunYJNckOTXJDm3bbsAt\nPZutYYKwSbI0yfIky9evXz+gqiVJQwuPJI8CzgXeWlW/Bk4GHg8sBtYCHxpbdZzNa7x9VtUpVbWk\nqpbMmzdvAFVLkmBI4ZHkoTTB8Zmq+iJAVd1WVfdU1b3Ax9nYNbUG2L1n8/nArTNZryTp/oZxtVWA\nTwLXV9WHe9p37VntpcDK9v0y4LAkD0uyB7AIuHym6pUkbW4YV1s9E3gNcG2Sq9u2vwEOT7KYpktq\nNfAGgKq6Lsk5wA9prtQ6xiutJGm4Zjw8qurfGH8c44JJtjkBOGFgRUmSOvEOc0lSZ4aHJKkzw0OS\n1JnhIUnqzPCQJHVmeEiSOjM8JEmdGR6SpM4MD0lSZ4aHJKkzw0OS1JnhIUnqzPCQJHVmeEiSOjM8\nJEmdGR6SpM4MD0lSZ4aHJKkzw0OS1JnhIUnqbNaER5KDktyQZFWS44ZdjyTNZVsPu4B+JNkK+Ajw\np8Aa4Ioky6rqh8Osa+FxX53W/a1+34umdX+SNCizIjyA/YBVVXUTQJKzgUOAoYbHdJvuMAIDSdJg\nzJbw2A24pWd5DfD0IdUyqwwikOaa6Q7gfv83GdZxuxx7WL+Lhm+2hEfGaavNVkqWAkvbxd8kuWGK\nx9sZ+PkUtx0m6x6AvH/CjwZa9yTHfaC2WPd0H3ua9jfS35NJzKa6H9fvirMlPNYAu/cszwdu3XSl\nqjoFOOWBHizJ8qpa8kD3M9Ose2ZZ98yy7tEyW662ugJYlGSPJNsAhwHLhlyTJM1Zs+LMo6o2JDkW\nuBDYCji1qq4bclmSNGfNivAAqKoLgAtm6HAPuOtrSKx7Zln3zLLuEZKqzcadJUma1GwZ85AkjRDD\no8dsmgIlyalJ1iVZ2dO2Y5KLk9zY/txhmDVuKsnuSb6Z5Pok1yV5S9s+0nUDJHl4ksuT/KCt/X+1\n7Xsk+X5b++faCzpGSpKtklyV5Cvt8sjXDJBkdZJrk1ydZHnbNhu+K9sn+UKSH7Xf9WfMhrq7Mjxa\nPVOgvBDYGzg8yd7DrWpSpwEHbdJ2HHBJVS0CLmmXR8kG4O1V9URgf+CY9t941OsGuBt4blU9FVgM\nHJRkf+D9wIlt7b8Ejh5ijRN5C3B9z/JsqHnMn1TV4p5LXWfDd+WfgK9X1R8BT6X5t58NdXdTVb6a\ncZ9nABf2LB8PHD/surZQ80JgZc/yDcCu7ftdgRuGXeMW6v8SzXxls63uRwJX0sxy8HNg6/G+Q6Pw\norkn6hLgucBXaG64Hemae2pfDey8SdtIf1eA7YCbaceTZ0vdU3l55rHReFOg7DakWqbqMVW1FqD9\nucuQ65lQkoXA04DvM0vqbrt/rgbWARcDPwZ+VVUb2lVG8Tvzj8D/AO5tl3di9GseU8BFSVa0s0fA\n6H9X9gTWA59quwo/kWRbRr/uzgyPjfqaAkUPXJJHAecCb62qXw+7nn5V1T1VtZjmr/n9gCeOt9rM\nVjWxJC8G1lXVit7mcVYdmZo38cyq2oemK/mYJM8ZdkF92BrYBzi5qp4G/JYHQxfVOAyPjfqaAmXE\n3ZZkV4D257oh17OZJA+lCY7PVNUX2+aRr7tXVf0K+BbNuM32Scbulxq178wzgZckWQ2cTdN19Y+M\nds33qapb25/rgPNoAnvUvytrgDVV9f12+Qs0YTLqdXdmeGz0YJgCZRlwZPv+SJoxhZGRJMAngeur\n6sM9H4103QBJ5iXZvn3/COD5NAOh3wRe3q42UrVX1fFVNb+qFtJ8n/+1ql7NCNc8Jsm2SR499h44\nEFjJiH9XqupnwC1JntA2PY/m0REjXfdUeJNgjyQH0/xlNjYFyglDLmlCSc4CDqCZsfM24N3A+cA5\nwALgJ8ArquoXw6pxU0meBXwXuJaNffB/QzPuMbJ1AyR5CnA6zXfjIcA5VfXeJHvS/FW/I3AVcERV\n3T28SseX5ADgHVX14tlQc1vjee3i1sBnq+qEJDsx+t+VxcAngG2Am4DX035nGOG6uzI8JEmd2W0l\nSerM8JAkdWZ4SJI6MzwkSZ0ZHpKkzgwPqUeSbyUZ+POmk/xVO+PqZ0ahHqmrWfMkQWnUJdm6Z86o\nLXkz8PyqWjPImsZ0rE3aIs88NOskWdj+1f7x9tkaF7V3fd/vL/UkO7dTc5DkdUnOb5+lsDrJsUne\n1k5ed1mSHXsO8Zr2GRIrk+zXbr9tmmeoXNFuc0jPfpcl+Vea2Ws3rfVt7X5WJnlr2/ZRmgn0vpbk\nv2+y/iOSnN3+fucBj+j57MAklya5Msnn2znCSHJw++yIFUlOysbndrwnyZlJvgec2U7s+IH2d7gm\nyRt69v3XPe1jzyrZNslX0zzDZGWSVz2g/+H0oOKZh2arRcDhVfUXSc4B/hz49Ba2eTLNTL4PB1YB\n76yqpyU5EXgtzewCAI+sqsXtRHynttv9Lc30Hke105RcnuQb7fr7AE/Z9I7hJPvS3F38dJoJCb+f\n5NtV9cYkB9E8q+Lnm9T4JuCuqnpie1f7le2+dgbeRXO28tsk7wTeluQfgI8Bz6mqm9uZB3rtDTyr\nqn6XZmbaO6rqvyV5GPC9JBe1/5aLaOaOCrCs/d3nAbdW1YvaGv5gC/++mkMMD81WN1fV1e37FTTP\nNtmSb1bVncCdSe4Avty2Xws8pWe9swCq6jtJtmvD4kCaSQbf0a7zcJqpJgAunmCqiWcB51XVbwGS\nfBF4Ns2UIBN5DnBSe/xrklzTtu9PEwTfa6YIYxvgUuCPgJuq6uae2pf27G9ZVf2ufX8g8JQkY/Na\n/QFNaBzYvsbqelTb/l3gg0neD3ylqr47Sd2aYwwPzVa9czHdw8bunQ1s7I59+CTb3NuzfC/3///C\npnP2FM1f5H9eVTf0fpDk6TTTbo9nvOnP+zHenEGhCanDNzn+07awr97aAvxlVV24yT5eAPyfqvrY\nZgdtzp4OBv4+ySVV9d5+fgE9+DnmoQeb1cC+7fuXT7LeZF4F903keEdV3QFcCPxlOzNwP//RBvgO\ncGiSR7Yzw76U5q/5LW3z6vYYT2bjGdFlwDOT/GH72SOT7AX8CNgzzcO17qt9AhcCb0ozLT5J9mrr\nuhA4qmcMZbckuyR5LE0X2qeBD9B0z0mAZx568PkgcE7bv//VKe7jP5NcBTwUOKpt+zuaMZFr2gBZ\nDbx4sp1U1ZVJTgMub5s+UVWTdVkBnEzzFLrraaZ8X9Hua32S1wFnteMVAO+qqn9P8mbg60l+S/No\ngYl8gqZ778r2d1gPHFpVFyV5InBpm42/AY4A/hD4QJJ7gd/TjMdIgLPqSrNekkdV1W/aQPgIcGNV\nnTjsuvTgZreVNPv9RZpnq19HMwi+2diFNN0885AkdeaZhySpM8NDktSZ4SFJ6szwkCR1ZnhIkjoz\nPCRJnf1/dlEE2eRW+MMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fba80320290>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plt.hist(miRNAs_nodes_degrees[1].values())\n",
    "plt.hist(targets_nodes_degrees[0].values(), bins=30)\n",
    "plt.ylabel('counts of nodes')\n",
    "plt.xlabel('number of degrees')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}